lambda-buffers-compiler/src/LambdaBuffers/Compiler/KindCheck/Inference.hs

```diff
@@ -0,0 +1,413 @@
+runDerive'' :: Type -> Either InferErr (Derivation, [Constraint])
+runDerive'' = runDerive testingContext
+
+-- | Run derivation - throw error.
+runDerive' :: Type -> (Derivation, [Constraint])
+runDerive' t = either (error.show) id $ runDerive'' t
+
+-- | Run derivation builder - not unified yet.
+runDerive :: Context -> Type -> Either InferErr (Derivation, [Constraint])
+runDerive ctx t = run $ runError $ runWriter $ evalState (DC atoms) $ runReader ctx (derive t)
+
+infer :: Context -> Type -> Either InferErr Kind
+infer ctx t = do
+  (d, c) <- runDerive (defTerms <> ctx) t
+  s <- runUnify' c
+  let res = foldl (flip substitute) d s
+  pure $ res ^. topKind
+  where
+    defTerms =
+      mempty
+        & context
+          .~ [ ("Either", Type :->: Type :->: Type)
+             , ("()", Type)
+             , ("Void", Type)
+             , ("(,)", Type :->: Type :->: Type)
+             ]
+
+--------------------------------------------------------------------------------
+-- Implementation
+
+-- | Creates the derivation
+derive :: Type -> Derive Derivation
+derive x = do
+  c <- ask
+  case x of
+    Var at -> do
+      v <- getBinding at
+      pure $ Axiom $ Judgement (c, x, v)
+    App t1 t2 -> do
+      d1 <- derive t1
+      d2 <- derive t2
+      let ty1 = d1 ^. topKind
+          ty2 = d2 ^. topKind
+      v <- KVar <$> fresh
+      tell [Constraint (ty1, ty2 :->: v)]
+      pure $ Application (Judgement (c, x, v)) d1 d2
+    Abs v t -> do
+      newTy <- KVar <$> fresh
+      d <- local (\(Context ctx addC) -> Context ctx $ (v, newTy) : addC) (derive t)
+      let ty = d ^. topKind
+      freshT <- KVar <$> fresh
+      tell [Constraint (freshT, newTy :->: ty)]
+      pure $ Abstraction (Judgement (c, x, freshT)) d
+  where
+    fresh :: Derive Atom
+    fresh = do
+      (DC vs) <- get
+      case vs of
+        a : as -> put (DC as) >> pure a
+        [] -> throwError $ ImpossibleErr "End of infinite stream"
+
+{- | Gets the binding from the context - if the variable is not bound throw an
+ error.
+-}
+getBinding :: Atom -> Derive Kind
+getBinding t = do
+  ctx <- asks getAllContext
+  case t `lookup` ctx of
+    Just x -> pure x
+    Nothing -> throwError $ UnboundTermErr $ show (pretty t)
+
+-- | Gets kind from a derivation.
+topKind :: Getter Derivation Kind
+topKind = to f
+  where
+    f = \case
+      Axiom (Judgement (_, _, k)) -> k
+      Abstraction (Judgement (_, _, k)) _ -> k
+      Application (Judgement (_, _, k)) _ _ -> k
+
+-- | Unification monad.
+type Unifier a = forall effs. Member (Error InferErr) effs => Eff effs a
+
+-- | Gets the variables of a type.
+getVariables :: Kind -> [Atom]
+getVariables = \case
+  Type -> mempty
+  x :->: y -> getVariables x <> getVariables y
+  KVar x -> [x]
+
+--------------------------------------------------------------------------------
+-- Unification
+
+-- | Unifies constraints and creates substitutions.
+unify :: [Constraint] -> Unifier [Substitution]
```

Comment:
Consider unification-fd library and check #9

Reply:
I'll consider this in the next version.